sensor/common/networkflow/manager/manager_impl.go

```diff
@@ -818,59 +960,107 @@ func (m *networkFlowManager) enrichProcessListening(ep *containerEndpoint, statu
+		// we cannot remove historical endpoints immediately. We must first make sure that network graph & PLOP have used
+		// the enrichment result. Because the `status.used` is assumed for the network graph & historical endpoints,
+		// we only check whether the enrichment result has been consumed by PLOP.
+		if status.historical {
+			return status.usedProcess
+		}
+		// For non-historical endpoints, we check whether both features already consumed the enrichment result.
+		return status.usedProcess && status.used && endpointClosed
+	}
+	// The `status.used` is assumed for the network graph & historical endpoints, because this function is executed
+	// after the enrichment finishes. We do not check `endpointClosed`, because in this case the Sensor makes the
+	// decision to close the endpoint (the close-message from Collector may never arrive). Sensor can make such decision
+	// because the container ID associated with this endpoint has been deleted already and we keep the entry in history
+	// only for a short time period.
+	if status.historical {
+		return true
+	}
+	return status.used && endpointClosed
 }
 
 func (m *networkFlowManager) enrichHostContainerEndpoints(hostConns *hostConnections, enrichedEndpoints map[containerEndpointIndicator]timestamp.MicroTS) {
 	hostConns.mutex.Lock()
 	defer hostConns.mutex.Unlock()
 
-	prevSize := len(hostConns.endpoints)
+	flowMetrics.FlowEnrichments.WithLabelValues("endpoint").Add(float64(len(hostConns.endpoints)))
 	for ep, status := range hostConns.endpoints {
 		m.enrichContainerEndpoint(&ep, status, enrichedEndpoints)
-		// If processes listening on ports is enabled, it has to be used there as well before being deleted.
-		used := status.used && (status.usedProcess || !env.ProcessesListeningOnPort.BooleanSetting())
-		if status.rotten || (used && status.lastSeen != timestamp.InfiniteFuture) {
-			// endpoints that are no longer active and have already been used can be deleted.
+		if shallRemoveEndpoint(status) {
 			delete(hostConns.endpoints, ep)
+			flowMetrics.HostConnectionsOperations.WithLabelValues("remove", "endpoints").Inc()
 		}
 	}
-	flowMetrics.HostEndpointsRemoved.Add(float64(prevSize - len(hostConns.endpoints)))
 }
 
 func (m *networkFlowManager) enrichProcessesListening(hostConns *hostConnections, processesListening map[processListeningIndicator]timestamp.MicroTS) {
 	hostConns.mutex.Lock()
 	defer hostConns.mutex.Unlock()
 
-	prevSize := len(hostConns.endpoints)
 	for ep, status := range hostConns.endpoints {
 		if ep.processKey == emptyProcessInfo {
-			// No way to update a process if the data isn't there
+			flowMetrics.IncHostProcessesEnrichmentEvents(
+				"N/A", "skip-enrichment", "N/A", "emptyProcessInfo",
+				status.lastSeen != timestamp.InfiniteFuture, status.rotten, false, false)
+			// Impossible to update a process if the data isn't there -> no chance for enrichment in a retry.
+			delete(hostConns.endpoints, ep)
```

Comment:
~Would it be possible to set here the `status.usedProcess` to true yield the same result? I wonder if it makes sense to enrich the endpoint in `enrichContainerEndpoint` even if the `processKey` is empty.~ 
Never mind, I think this won't be a problem because `enrichContainerEndpoint` runs before `enrichProcessesListening`.

Reply:
Yeah, this is a bit confusing. I want to propose a refactor for that part and I have a branch for that already in the stack: https://github.com/stackrox/stackrox/pull/14880

Reply:
Yeah, I messed-up a rebase for #14880 (I'm using git-spice for branch stacking and it didn't work well for that rebase) and decided to close it. I plan to open another PR with the same goal after we are close to merge this branch to avoid huge rebasings.